onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -31,58 +31,54 @@ namespace onnxruntime {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Internal error. The preallocated buffer is too small. Requires ",
+                             p_tensor->SizeInBytes(), ", Got ", m->GetLen());
+    }
+  } else {
+    // tensor constructor should give us enough buffer size based on type and shape
+    p_tensor = onnxruntime::make_unique<Tensor>(type, tensor_shape, alloc);
   }
 
-  // deserialize to CPU first for non-CPU allocator, then alloc and copy
-  size_t cpu_tensor_length;
-  ORT_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<0>(tensor_proto, &cpu_tensor_length));
-  if (m.GetLen() < cpu_tensor_length) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Internal error. The preallocated buffer is too small. Requires ",
-                           cpu_tensor_length, ", Got ", m.GetLen());
-  }
+  if (strcmp(p_tensor->Location().name, CPU) == 0) {
+    // deserialize directly to CPU tensor
+    ORT_RETURN_IF_ERROR(utils::TensorProtoToTensor(env, proto_path.c_str(), tensor_proto, *p_tensor));
+  } else {
+    // non-cpu tensor
+    if (tensor_proto.data_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "string tensor is not supported for copying between allocators");
+    }
 
-  std::unique_ptr<char[]> data(new char[cpu_tensor_length]);
-  std::unique_ptr<Tensor> p_tensor;
-  OrtValue tmp_ort_value;
-  OrtCallback d;
-  ORT_RETURN_IF_ERROR(utils::TensorProtoToMLValue(env, proto_path.c_str(), tensor_proto,
-                                                  MemBuffer(data.get(), cpu_tensor_length, default_cpu_memory_info),
-                                                  tmp_ort_value, d));
-
-  const Tensor& p_deserialize_tensor = tmp_ort_value.Get<Tensor>();
-
-  p_tensor = onnxruntime::make_unique<Tensor>(p_deserialize_tensor.DataType(), p_deserialize_tensor.Shape(), m.GetBuffer(),
-                                              m.GetAllocInfo());
-  // TODO: does this function work for string tensor?
-  Status copy_status = data_transfer_mgr.CopyTensor(p_deserialize_tensor, *p_tensor);
-  if (d.f) d.f(d.param);
-
-  if (!copy_status.IsOK()) {
-    if (copy_status.ErrorMessage().empty()) {
-      // The windows execution provider does not return any error message today for CopyTensor since it is
-      // not implemented yet. That's the reason we're adding our own error message so that we can debug better.
-      return Status(copy_status.Category(), copy_status.Code(),
-                    "Failed to copy tensor to " + p_tensor->Location().ToString());
+    // deserialize to CPU first for non-CPU allocator, then copy
+    std::unique_ptr<Tensor> p_deserialize_tensor = onnxruntime::make_unique<Tensor>(type, tensor_shape, default_cpu_alloc);
+    ORT_RETURN_IF_ERROR(utils::TensorProtoToTensor(env, proto_path.c_str(), tensor_proto, *p_deserialize_tensor));
+    // TODO!! Need a temp buffer allocator for non-escape buffers that maybe too big for stack allocation.
```

Comment:
> // TODO!! Need a temp buffer allocator for non-escape buffers that maybe too big for stack allocation. [](start = 3, length = 103)

What is this comment for?

Reply:
In memory management, it's usually a good idea to group objects with similar life time together. Here this Tensor has a very short life, but we are mixing it with other tensors that potential last a very long time. Putting this Tensor on an special allocator just for temp objs would reduce fragmentation in the main allocator.